Reduxify the user-settings module (/me/settings endpoint)

jsnajdr opened the pull request:
This module has API almost identical to `client/lib/user-settings`, but
it's Redux - there's state and reducers, and the API are selector and
action creator functions, not object methods.

It's still a work in progress - I'm just requesting feedback now, not ready for merge.

My next steps would be writing a test suite and then replacing the `lib/user-settings` usage at least at one place. It will help a lot with moving the language picker project (PR #12544) forward.

The `lib/user-settings` module should eventually be removed once all usage sites are updated.

This module is very closely related to `client/state/preferences`. That module also works with the `wpcom.me().settings()`, but cherry-picks just the `calypso_preferences` property and ignores the rest. Maybe both modules should be merged into one?

`client/state/preferences` has some additional features:
- supports default values when the network is not available
- validates the data against a schema
- has support for setting a preference only locally, with no intent to ever save it to the server. My user settings module only supports unsaved values - the implementation is very similar, but the semantics are different

There's also a second module for preferences: `client/lib/preferences`. That one is a Flux store, with functionality seemingly identical to the `client/state` one, with one extra feature - it supports storing and loading the preferences to local storage.

I'm looking for feedback whether the `user-settings` and `preferences` modules are worth merging. There is not a straightforward 1:1 relation between them. @ehg, what do you think?

I'm also wondering whether using the `client/state/data-layer` for accessing the `wpcom` API would bring any direct benefit here. I'm not very familiar with the `data-layer` at this moment. Can @dmsnell help?

dmsnell commented on client/state/user-settings/README.md:
````diff
@@ -0,0 +1,23 @@
+
+1. Render `QueryUserSettings` from `components/data/query-user-settings`
+2. Connect your component specifying proper setting names:
+```es6
````
should this not be `js` here?

anyways, although we all know and love it as ES6 it's just JavaScript and the official name of the release is ES2015.

spen replied:
A few months back I updated a few (maybe a lot) of the readmes to `es6` syntax where they previously had no highlight tag - At the time some ES6+ specific syntax wasn't supported with the JS tag.
I'm pretty sure that at some point either Github or PrismJS (can't remember which) stopped highlighting `ES6` tags but at the same time `JS` was updated to cover the `es6` syntax.
I've just checked and [devdocs highlights es6](https://wpcalypso.wordpress.com/devdocs/client/lib/devices/README.md?term=es6)  properly, as does Github - but I think I had to alias `es6` to `js` for the devdocs syntax highlighting to work.
My memory is fuzzy on this though - it was a little while ago now!

`es6` works but I agree it should be `js` mostly just for future-proofing.

dmsnell commented on client/state/user-settings/actions.js:
```diff
@@ -0,0 +1,171 @@
+
+		// Let the caller know whether the fetch was successful or not
+		return fetchRequest;
+	};
```
I'll highly encourage a few things with this. Number one is to eliminate introducing a new Redux thunk action. Number two is to try writing this from inside the data layer. That _would_ require figuring out the actual API endpoint path and other settings specific to `wpcom.me().settings().get()` because the data layer prefers declarative HTTP request actions over direct stateful calling and promises..

In this process we can get rid of the abundance of extra meta actions ending in `_SUCCESS` and `_FAILURE`; also to change the nomenclature of `FETCH` to `REQUEST` and the semantics of `SETTINGS_RECEIVE` to something more meaningful in terms of Calypso intentions, like `SETTINGS_UPDATE`.

dmsnell commented on client/state/user-settings/actions.js:
```diff
@@ -0,0 +1,171 @@
+export function receiveUserSettings( values ) {
+	return {
+		type: USER_SETTINGS_RECEIVE,
+		values
```
notice here how we have hidden any clues about the shape and structure of `values`

I hoped that by looking into the action creator I could discover what is expected there but it left me without much help. what can we do in cases like this? well we could provide some examples in the JSDoc of what this could look like, describe in the `@param` tag, or explicitly show the shape by using more split-up values in the action itself

dmsnell commented on client/state/user-settings/reducer.js:
```diff
@@ -0,0 +1,82 @@
+		display_name: data.display_name && decodeEntities( data.display_name ),
+		description: data.description && decodeEntities( data.description ),
+		user_URL: data.user_URL && decodeEntities( data.user_URL )
+	};
```
awesome job normalizing the data from the API!

unfortunately, remember that this _isn't_ technically data coming in from the API…what's that? It's actually data coming in from the dispatcher, which in most cases happens to come originally from the API.

it would be wonderful if the encoded information never ever entered the Calypso world by normalizing where we get the network request. this of course has its spot in the data layer files where the entire mapping stays in one easily-discoverable spot.

for example, if I want to test some code with a given `display_name` then I'd just as soon open the dev tools and dispatch an action with a predefined name but I don't want to have to know that I need to first encode that string because the API does…

dmsnell commented on client/state/user-settings/reducer.js:
```diff
@@ -0,0 +1,82 @@
+	[ USER_SETTINGS_RECEIVE ]: ( state, { values } ) => {
+		return merge( {}, state || {}, decodeUserSettingsEntities( values ) );
+	}
+} );
```
I don't see much value gained in using `createReducer` here since we have a very simple reducer which doesn't handle serialization or memoization. Is there a reason I'm missing why it's used?

dmsnell replied:
You mentioned that there is overlap in the settings here as with other reducers? Are we duplicating the same state here where instead we could reuse the other state (a superset of the settings we need here)?

jsnajdr replied:
> You mentioned that there is overlap in the settings here as with other reducers? Are we duplicating the same state here where instead we could reuse the other state (a superset of the settings we need here)?

Actually, it's the other way around: There is a `client/state/preferences` module that fetches the `me/settings` endpoint and stores a *subset* of the settings, namely the `calypso_preferences` key. The value of that key is a JSON structure that is stored in `state.preferences`.

The `client/state/preferences` module could be merged into this one after it's finished.

jsnajdr replied:
> I don't see much value gained in using `createReducer` here since we have a very simple reducer which doesn't handle serialization or memoization. Is there a reason I'm missing why it's used?

I use `createReducer` because it provides a nice API where I pass a map `actionType -> handler` instead of creating a big `switch` statement. And also because it seems to be used everywhere else. Or is the preferred style to use the good old `switch`?

The default implementations for `SERIALIZE` and `DESERIALIZE` return the initial state. So, `DESERIALIZE` does the state hydration, and `SERIALIZE` kind of annihilates the state (?). I don't quite understand what these actions are supposed to do, except schema validation.

I'm not sure what "memoization" means when talking about a reducer.

dmsnell replied:
> I use createReducer because it provides a nice API where I pass a map actionType -> handler instead of creating a big switch statement.

as a first observation here, we wouldn't be dealing with any big `switch` statement because we're only listening to a single action type. all of this work is just complicating what could be a single simple conditional.

> it seems to be used everywhere else. Or is the preferred style to use the good old switch?

well, the premise is a false dichotomy but yes it's often used but not everywhere. I encourage people not to jump to `createReducer` unless it brings some specific benefit they want because it adds overhead and complexity to the code. more likely than not, `createReducer` doesn't do what we think it does, which is difficult for maintaining code that uses it. it's also _not_ used everywhere, as plenty of reducers are simply built to their needs.

```js
export const settings = ( state = null, { type, values } ) =>
	USER_SETTINGS_RECEIVE === type
		? { ...state, ...decodeUserSettingsEntities( values ) }
		: state;
```

dmsnell commented on client/state/user-settings/reducer.js:
```diff
@@ -0,0 +1,82 @@
+	[ USER_SETTINGS_FETCH_SUCCESS ]: () => false,
+	[ USER_SETTINGS_FETCH_FAILURE ]: () => false,
+	[ USER_SETTINGS_FETCH ]: () => true,
+} );
```
in the data layer we could almost certainly get away without this. one of the goals of that layer is to remove the need for any kind of `isFetching` boolean states.

dmsnell commented on client/state/user-settings/reducer.js:
```diff
@@ -0,0 +1,82 @@
+
+export const initialized = createReducer( false, {
+	[ USER_SETTINGS_FETCH_SUCCESS ]: () => true
+} );
```
we probably don't need this either. `null` is a useful value to indicate that we have no data yet and we can use that to infer whether or not the data has already come in

dmsnell commented on client/state/user-settings/selectors.js:
```diff
@@ -0,0 +1,104 @@
+		setting = settings[ settingName ];
+	}
+
+	return setting;
```
we have no need to mutate `setting` here

```js
return settings && settings[ settingName ] !== undefined
	? settings[ settingName ]
	: null;

// or more simply
return get( settings, 'settingName', null )
```

dmsnell commented on client/state/user-settings/selectors.js:
```diff
@@ -0,0 +1,104 @@
+export function hasUnsavedUserSettings( state ) {
+	const { unsavedSettings } = state.userSettings;
+	return ! isEmpty( unsavedSettings );
+}
```
please note that all new selectors should live one per file in the `state/selectors` directory

dmsnell commented:
Nice contributions @jsnajdr!

There are some very high-level suggestions I might have with this PR before trying to dive too deeply into the details. There's a lot we can trim from the reducer functions both in their definitions but also in how many we need. I think there's quite a bit of overlap in some of the reducers - meaning there's information we could infer from other data instead of storing it separately.

Further, I do encourage looking into the data and HTTP layers. They should be able to shave off quite a bit from this PR while providing the reliability we want. I'd be happy to work with you on this and answer any questions you might have.

Finally, check and see if we can't actually use code that already exists for our purposes. I somewhat suspect that there might be a way to reuse the user settings in Calypso with just a little augmentation.

dmsnell commented on client/state/user-settings/reducer.js:
```diff
@@ -0,0 +1,82 @@
+	[ USER_SETTINGS_UNSAVED_REMOVE ]: ( state, { settingName } ) => {
+		return omit( state, settingName );
+	}
+} );
```
since the comment was brought up we still have no need to abstract up to `createReducer` when dealing with lists of action types

```js
export const unsavedSettings = ( state = null, { type, savedSettingNames, settingName, value } ) => get( {
	[ USER_SETTINGS_SAVE_SUCCESS ]: () => savedSettingNames ? omit( state, savedSettingNames ) : {},
	[ USER_SETTINGS_UNSAVED_SET ]: () => state[ settingName ] === value ? state : { ...state, [ settingName ]: value },
	[ USER_SETTINGS_UNSAVED_REMOVE ]: () => omit( state, settingName ),
}, action, identity )( state );
```

jsnajdr replied:
This code is almost identical to what `createReducer` would do, but has some performance issues - the map object from action types to handlers will get re-created on every reducer invocation, and then garbage-collected on return. So I'd prefer using the `createReducer` helper here, if you don't mind.

dmsnell replied:
sure, but it's easy enough to separate so they don't get recreated every time. I was just pointing out that `createReducer()` carries its own overhead: both performance wise and in hidden behaviors.

jsnajdr commented:
@dmsnell I just pushed a new version of the PR where I changed the following:

- use the `data-layer` to send HTTP requests to the `/me/settings` endpoint. There's no more a need for an `isFetchingUserSettings` state and a zoo of network-related actions like `USER_SETTINGS_REQUEST_SUCCESS` of `USER_SETTINGS_SAVE_FAILURE`
- the actions dispatched from the data layer into the Calypso app don't have any network-semantics any more.
- I rewrote the `user-settings` reducers to use `switch` statements. I'd like to rewrite them back into more functional form, and I'd prefer `createReducer` - I'm posting a separate comment about this issue.

I'll proceed with writing a test suite for the new functionality, and with moving the selectors to the right place - one per file in `state/selectors`.

Let me know what you think!

jsnajdr commented:
So my `user-settings` Redux module is ready for review! it maintains the user settings data in a `userSettings` state slice, and uses the `data-layer` and `wpcom-http` handlers to fetch and save the settings over the network.

There is a test suite for the actions and reducer, for the data layer, and also for the selectors.

The last piece of work I need to finish is to move the selectors to the `client/state/selectors` directory in a one-per-file format.

jsnajdr commented:
Update: moved selectors to `client/state/selectors`, one per file.

dmsnell commented on client/state/data-layer/wpcom/user-settings/index.js:
```diff
@@ -0,0 +1,83 @@
+/**
```
we're trying to map the filenames 1:1 with the API endpoints to make this kind of thing discoverable and to make the mapping clear. I would anticipate this file being called `state/data-lyer/wpcom/me/settings/index.js` based on the path in the API call

jsnajdr replied:
OK, I'll move this to the right directory.

dmsnell commented on client/state/data-layer/wpcom/user-settings/index.js:
```diff
@@ -0,0 +1,83 @@
+	};
+
+	return merge( {}, apiResponse, decodedValues );
+}
```
👍 

also, I'm pretty sure we don't have the liberty here to rename `display_name` > `displayName`. is that right?

jsnajdr replied:
We'd have to be careful doing the inverse transformation when sending a request to server, and change all the places where the setting names are used. And all that for very little benefit.

dmsnell replied:
> And all that for very little benefit.

debatable :wink: we also write all this in JS instead of rendering server-generated HTML for very little benefit…

my main point was a question if we're already depending on these names in other places in Calypso. I consider it _a good thing_ to perform the mapping on ingress and egress which prevents the tight coupling of the API with Calypso and encourages us to do actual validation and explicit data transformation

dmsnell commented on client/state/data-layer/wpcom/user-settings/index.js:
```diff
@@ -0,0 +1,83 @@
+		user_URL: apiResponse.user_URL && decodeEntities( apiResponse.user_URL )
+	};
+
+	return merge( {}, apiResponse, decodedValues );
```
any particular preference for using `merge()` here over `{ ...apiResponse, ...decodedValues }`?

jsnajdr replied:
`merge()` ignores keys with `undefined` values. `decodedValues` will have keys like this if a setting wasn't present in the `me/settings` response. And that can happen - if you save just one setting to `POST me/settings`, the response will include only that one changed setting.

dmsnell replied:
that seems like a detail that warrants a comment. the need for the behavior isn't at all obvious unless you know it needs to be there. could you add as much above this line?

dmsnell commented on client/state/data-layer/wpcom/user-settings/index.js:
```diff
@@ -0,0 +1,83 @@
+	dispatch( updateUserSettings( fromApi( data ) ) );
+};
+
+export const receiveRequestError = noop;
```
maybe for now we could simply leave this out or pass in the `noop` in the `dispatchRequest()` call? am I missing some reason to leave it in here as an exported function?

jsnajdr replied:
Good point. I created the exports because I expected I would need them in the tests, but then I didn't. I'll remove the exports and pass `noop` directly to `dispatchRequest`.

dmsnell commented on client/state/data-layer/wpcom/user-settings/index.js:
```diff
@@ -0,0 +1,83 @@
+ */
+export function saveUserSettings( { dispatch, getState }, action, next ) {
+	const { settingsOverride } = action;
+	const settings = settingsOverride || getState().userSettings.unsavedSettings;
```
if I understand these lines properly, what we're saying is that _if_ the action has any settings overrides then we completely ignore the unsaved settings, otherwise we save the unsaved settings?

would not merging them also make sense?

```js
const body = {
	...getUnsavedSettings(),
	...action.settingsOverride,
};
```

note that we should probably be calling a selector for the unsaved settings. it's not wrong as-is, and it's not for performance reasons, just for unhooking the state data from the app so we can preserve flexibility in how we store the state data.

jsnajdr replied:
There are two ways how to work with this module in the UI:
1. Present a form where the user can change one or more values and then hit "Submit". In this case, we dispatch `updateUserSetting` on every field change and one `saveUserSettings` without parameters on submit. The intermediate unsaved state is stored in `unsavedSettings`.
2. Present a control that changes certain setting immediately without any intermediate workflow. For example, we will have a one-click language switcher in the sidebar soon. Such a control will directly dispatch `saveUserSettings` with just one changed value. And it will not disrupt the unsaved form that might me displayed on the same screen.

It might not be ideal API, but it's how the current `client/lib/user-settings` module works and it works well in all places where it's used in Calypso.

I'll use the `getUnsavedSettings` selector. Also, I'll rename the `updateUserSetting` action to `setUserSetting`. When renaming the `receive*` actions to `update*`, I didn't notice that I created a confusing name conflict - the actions for altering a setting and for receiving settings from server both use the `update` verb.

dmsnell replied:
thanks for the explanation. it still seems unclear to me what the intention of this code is. did I correctly infer that in my comment about?

```js
if hasActionOverrides():
	saveOnlyOverrides()
else:
	saveOnlyUnsavedChanges()
```

jsnajdr replied:
> thanks for the explanation. it still seems unclear to me what the intention of this code is. did I correctly infer that in my comment about?

Yes, that's exactly right.

I'm adding a new `getUnsavedUserSettings` selector and using it instead of directly accessing the state.

dmsnell commented on client/state/data-layer/wpcom/user-settings/index.js:
```diff
@@ -0,0 +1,83 @@
+
+export const requestUserSettings = ( { dispatch }, action, next ) => {
+	dispatch( http( {
+		apiVersion: '1.1',
```
just a quick-check - this shouldn't be `v1.1` should it?

jsnajdr replied:
No, it's just `1.1`. The `v` will be added by the HTTP layer when constructing the request URL. By the way, if I don't specify the `apiVersion` here, the default value is `v1`, and the URL is then `.../vv1/...`. Maybe a bug?

dmsnell replied:
good catch. I'll check on that soon. I don't think the `apiVersion` should be optional but I'm also about to overhaul some of that to support `apiNamespace`

dmsnell commented on client/state/data-layer/wpcom/user-settings/index.js:
```diff
@@ -0,0 +1,83 @@
+	return next( action );
+}
+
+export const receiveSavedUserSettings = ( { dispatch }, { settingsOverride }, next, data ) => {
```
the name here I think could be more expressive in what it's doing. it's maybe not so much _receiving_ as it is _updating_ or _storing_. if we called this function without an API call being attached the nuance might look clearer:

```js
// in dev tools
receiveSavedUserSettings( store, settingsAction, store.dispatch, {
	… // mock settings data
} );
```

this wasn't _received_ any more than other direct function calls are _parameter receivers_

jsnajdr replied:
I believe that the purpose of the `receive` handlers really is to _receive_ things. They are invoked when a response from the server arrives, and their `data` parameter is in the wire format. They are clearly a part of the data layer.

The `receive` handler further dispatches one or more actions, and it's these actions that do the actual _updating_ or _storing_, not the handler itself.

If I called the `receiveSavedUserSettings` in the devtools REPL, the meaning would be: here is a JSON POST response we received from server, please do the work that needs to be done: display notices, update the store data, clear unsaved state on a form, redirect to a different page...

If I wanted to do just the store update from the REPL, I would directly dispatch the `updateUserSettings` action.

dmsnell replied:
there's no reason to block on the name here, but still it's obvious from the code that it's actually receiving the response but that doesn't really clue us in to its purpose. the reason I suggested a name change is so that the name can convey the intention of this function the same way we want to indicate the _why_ in a comment more than the _what_, since the _what_ is often obvious but the _why_ less so

dmsnell commented on client/state/data-layer/wpcom/user-settings/index.js:
```diff
@@ -0,0 +1,83 @@
+	dispatch( clearUnsavedUserSettings( settingsOverride ? keys( settingsOverride ) : null ) );
+
+	// TODO: Refetch the user data after saving user settings
+	// user.fetch();
```
why is this a TODO? is there a reason to not go ahead and issue the fetch here?

jsnajdr replied:
Just a forgotten TODO, will fix.

dmsnell commented on client/state/data-layer/wpcom/user-settings/index.js:
```diff
@@ -0,0 +1,83 @@
+	// user.fetch();
+};
+
+export const receiveSaveError = noop;
```
see above note on `receiveRequestError`

dmsnell commented on client/state/data-layer/wpcom/user-settings/index.js:
```diff
@@ -0,0 +1,83 @@
+export const dispatchUserSettingsSave = dispatchRequest( saveUserSettings, receiveSavedUserSettings, receiveSaveError );
+
+export default {
+	[ USER_SETTINGS_REQUEST ]: [ dispatchUserSettingsRequest ],
```
interesting… I'm a little curious is all why we have separate variables for these dispatchers vs simply issuing them in the export statement

jsnajdr replied:
I probably anticipated using these exports in unit tests... Didn't happen, so I'm merging them directly into the exported object.

dmsnell commented on client/state/data-layer/wpcom/user-settings/test/index.js:
```diff
@@ -0,0 +1,163 @@
+					description: 'foo & bar',
+					language: 'qix',
+				} ) );
+			} );
```
nice tests here! I appreciate how you are using _unit_ tests and letting the other layers (selectors, dispatcher) to test their own behaviors and leaving that coupling out of this file

dmsnell commented on client/state/selectors/get-original-user-setting.js:
```diff
@@ -0,0 +1,15 @@
+ * @return {*} setting key value
+ */
+export default function getOriginalUserSetting( state, settingName ) {
+	return get( state.userSettings.settings, [ settingName ], null );
```
seems a bit odd to use the safe-defaulting getter and then doing the nested property access in the _unsafe_ part of the function call…

```js
get( state, [ 'userSettings', 'settings', settingName ], null );
```

dmsnell commented on client/state/selectors/get-user-setting.js:
```diff
@@ -0,0 +1,20 @@
+			: settings[ settingName ];
+	}
+
+	return setting;
```
I'm a bit confused glossing over this nested conditional. I wonder if we could simplify the logic a bit…

```js
/**
 * Get user setting, prefer returning unsaved values first
 */
export const getUserSetting = ( state, settingName ) => {
	const unsaved = get( getUnsavedSettings( state ), settingName );

	// if the change exists in our unsaved queue, return it
	// otherwise try to get the saved value
	// fail to `null`
	return unsaved
		? unsaved
		: get( getUserSettings( state ), settingName, null );
}
```

jsnajdr replied:
The desired behavior is to return `null` also in case when there is an unsaved value, but the original value is not set. I'll use `_.get` to clean up the conditional a bit.

dmsnell commented on client/state/user-settings/README.md:
```diff
@@ -0,0 +1,23 @@
+		return bindActionCreators( {
+			saveLanguage: language => saveUserSettings( { language } ),
+		}, dispatch );
+	},
```
these are fine examples but also fairly verbose, maybe using arrow-function syntax for `connect()` could bring out the important details a bit more

```js
export default connect(
	state => ( { language: getUserSettings( state, 'language' ) } ),
	dispatch => ( { saveLanguage: language => dispatch( saveUserSettings( { language } ) ) } ),
)( Account );
```

dmsnell commented on client/state/user-settings/actions.js:
```diff
@@ -0,0 +1,116 @@
+	return {
+		type: USER_SETTINGS_REQUEST
+	};
+}
```
personal preference, but I like removing noisy syntax…

```js
export const fetchUserSettings = () => ( {
	type: USER_SETTINGS_REQUEST,
} );
```

jsnajdr replied:
Good idea, I like arrow functions, too :)

dmsnell commented on client/state/user-settings/actions.js:
```diff
@@ -0,0 +1,116 @@
+		}
+
+		return true;
+	};
```
is there a specific reason this still remains as an action thunk instead of being handled in the middleware?

jsnajdr replied:
This action thunk has nothing to do with the data layer. It's an action that sets the unsaved value of a setting. The logic depends on the current state that's being read with the `getState` function available to the thunk.

This is the price one needs to pay for having independent reducers that are composed with `combineReducers`. If the next state of state node `unsavedSettings` depends on state node `settings`, the action needs to be implemented as a thunk. The `unsavedSettings` reducer doesn't see the state of `settings`.

The `updateUserSetting` action name is confusing - I'll rename it to `setUserSetting`.

dmsnell replied:
it actually looks like this is only called from within the middleware, however, and making it a thunk has only moved the statefulness out of the middleware and into the action creator. this is a pattern I have a hard time accepting.

usually when I think that we need thunks or stateful action creators I recognize that I haven't fully considered alternatives often in the data structure itself

in this case since it's triggered from the middleware we have an easy way out - just pass in the old `settings` as an additional parameter

on the other hand, why do we have a specific `removeUnsavedUserSetting` action? do we actually need that or does the act of updating it imply that we can remove it? if the value is the same then an update should be idempotent and no real harm is done by going ahead and updating it. in this scenario we not only don't need to pass in the `settings` object and this becomes stateless, but we can have the unsaved settings reducer watch for the update action and remove the dirty bit

anyway, I'm brainstorming here because I don't know the behavioral goals as well as you do. introducing a new thunk is something I hope we can avoid entirely

jsnajdr replied:
No no, I think you are confused by the unfortunate name of the action. It should be (and will be) `setUserSetting`. It's intended to be used by UI components to actually change a setting value.

It has nothing to do with the `updateUserSettings` action that's used to update the state with data that came from the server.

The logic is:
- store the new value to `unsavedSettings`
- remove the value from unsaved settings when it's being reset to the original value. That's where `removeUnsavedUserSetting` comes into play. There is also a unit test for this case
- never store anything to `unsavedSettings` if it's not already in `settings`. The keys in `settings` define what values are supported. The unsupported ones are ignored. This applies both to this action thunk and also to the `getUserSetting` selector.

dmsnell commented on client/state/user-settings/reducer.js:
```diff
@@ -0,0 +1,54 @@
+		default:
+			return state;
+	}
+};
```
looks like we could simplify this reducer quite a bit…

```js
export const settings = ( state = false, { type, settingValues } ) =>
	USER_SETTINGS_UPDATE === type
		? { ...state, ...settingValues }
		: state;
```

also it seems odd to be storing the default value for _an object's state_ as `false`. might `null` be more appropriate here as the initial state?

jsnajdr replied:
Good point, I'm changing the default state value to `null`. The legacy `lib/user-settings` module used `false` and we have a good opportunity to make the change here.

I'd rather keep the reducer as a `switch` statement. It might be a bit more verbose, but it's 100% clear at first glance what the reducer does and which actions it handles. On the other hand, decoding the ternary operator takes some mental effort, at least for me :)

dmsnell replied:
> It might be a bit more verbose, but it's 100% clear at first glance what the reducer does and which actions it handles. On the other hand, decoding the ternary operator takes some mental effort, at least for me :)

easy for me hard for you
easy for you hard for me

jsnajdr commented:
> thanks for the updates! how do you feel about the change in code to use the http layer and data layer? I'm definitely interested in developer feedback.

At first, the data and HTTP layers were a bit intimidating and hard to understand, with all the abstractions and functional composition. I want to write a blogpost this week about my journey and what I learned.

There's one thing I don't know how to do yet and I'll need to do some research - and that's how to compose the actions dispatched to the data layer to form some larger "sagas".

For example, this is what I can do with action thunks that return promises:
```js
function saveSettings() {
  return function( dispatch ) {
    const request = wpcom.me().settings().update();
    request.then(
      data => dispatch( showSuccessNotice() )
      err => dispatch( showErrorNotice( err ) )
    );
    return request; // Give the caller a chance to handle the promise, too
  };
}

// In my component, I want to redirect after successful save
function mySaveSettings() {
  return function( dispatch ) {
    return dispatch( saveSettings() ).then( () => redirectTo( '/me/account' ) );
  };
}
```

How do I do this with the data layer? I don't know yet. With the data layer, the `saveSettings` action is a plain action, and dispatching it returns `undefined`.

The `http` action `wpcom-http` takes `onSuccess` and `onFailure` action parameters, maybe I should use a similar mechanism to customize handlers in the data layer?

dmsnell commented:
> How do I do this with the data layer? I don't know yet. With the data layer, the saveSettings action is a plain action, and dispatching it returns undefined.
> 
> The http action wpcom-http takes onSuccess and onFailure action parameters, maybe I should use a similar mechanism to customize handlers in the data layer?

yeah this is a consequence of us having stateful routing. we definitely have less direct coupling between action and response, but in cases where we need to we can still accommodate. we just need to ask ourselves _when_ do we want to redirect? why wait until the settings have come back from the server? why not just redirect `onClick`? if there is a problem saving the settings we can come back to it later.

this is actually touching on an important point about these inherently asynchronous systems - it's not linear. for example, the promise method breaks down if we queue the save when our internet connection drops or if we accidentally refresh the browser. on the other hand, if we've built a system considering that these events are asynchronously coupled then we can end up with something much more reliable and versatile.

honestly I just haven't seen any need before to depend on the promise being returned. in cases I've seen it used it was usually because of some weird state mangling going on in the component (a bigger issue) or for testing (because other parts of the system were built to be untestable otherwise)

dmsnell commented:
> At first, the data and HTTP layers were a bit intimidating and hard to understand, with all the abstractions and functional composition. I want to write a blogpost this week about my journey and what I learned.

thanks for the feedback! I'd love to read a post about it

jsnajdr commented:
@dmsnell Thanks for a very thorough review! I addressed the review comments and pushed a new version of the PR commits.

When adding a `user.fetch()` call to the `receiveSavedUserSettings` function, I had to resort to a dirty trick when importing the `lib/user` module. That module needs to be mocked in the unit test, and unfortunately not only in the `client/state/data-layer/wpcom/me/settings` tests. These handlers are merged into one giant module that is imported also in other tests that fail if the mock is not present. So I need to import the module at runtime only when the handler is actually _called_.

I'll think about a better name for the `receive*` handlers, I think I get your point.

jsnajdr commented:
Hi @dmsnell! I just pushed an updated version of this PR:
- I simplified the `settings` reducer, as you advised
- gave better names to the `receiveUserSettings` and `receiveSavedUserSettings` handlers

Let me please know if there are any remainining issues, or if we are ready to merge.

dmsnell commented on client/state/data-layer/wpcom/me/settings/test/index.js:
```diff
@@ -0,0 +1,166 @@
+
+				expect( dispatch ).to.have.been.calledTwice;
+				expect( dispatch ).to.have.been.calledWith( updateUserSettings( {
+					description: 'foo & bar'
```
should we be testing the entity decoding _and_ the save behavior in the same test? it seems like if some error with the decoded cropped up it would mistakenly fail this test.

maybe if we have a test for the decoding we could consider removing it from the other tests too?

jsnajdr replied:
Good point. I'll split the entity decoding test into a separate one. But it still will be test of the `storeFetchedUserSettings` and `finishUserSettingsSave` - I'll be testing if these functions call the `fromApi` utility function correctly, not just the `fromApi` function itself. Do I understand your point correctly?

dmsnell commented on client/state/selectors/is-two-step-sms-enabled.js:
```diff
@@ -0,0 +1,14 @@
+ * @return {Boolean} return true if two-step sms is enabled
+ */
+export default function isTwoStepSmsEnabled( state ) {
+	return get( state, 'userSettings.settings.two_step_sms_enabled', false );
```
is this a safe default? seems like _if_ someone has two-step enabled but for some reason there's an issue getting that value out of Redux then disabling two-step seems like weak security. granted, I suppose that any valid API calls would fairytales but the UX could suffer. of course, it could suffer if we ask for two-step and it doesn't exist…

any thoughts?

jsnajdr replied:
It is a safe default in a sense that the `isTwoStepSMSEnabled` method of the `lib/user-settings` objects behaves exactly the same way. So I'm keeping the status quo - it won't break the existing UI after it's migrated to the Redux version.

But returning `null` here is more correct -- the UI has a chance to distinguish between three states (unknown, true, false) if it decides to do so, and the default is still a falsy value. I'll change this.

dmsnell commented:
@jsnajdr I think I'm okay with this as it is now; at least I don't want to delay it any further, though I _did_ leave a couple of more comments. I would feel better if someone else did a review of the PR and if it weren't just me. I'll back out of the process for now though unless you poke me on it again though :smile:

jsnajdr commented:
Pushed two improvements:
- test the entity decoding of HTTP responses in a separate unit test
- selectors that return a boolean value (`isPendingEmailChange`, `isTwoStepEnabled`, ...) return `null` as a default if the value is unknown (i.e., not loaded yet).
